xtask/src/ast_src.rs

```diff
@@ -511,20 +1476,131 @@ pub(crate) const AST_SRC: AstSrc = AstSrc {
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/patterns.html#tuple-struct-patterns)
         struct TupleStructPat { Path, T!['('], args: [Pat], T![')'] }
+
+        /// Tuple pattern.
+        /// Note: this doesn't include tuple structs (see `TupleStructPat`)
+        ///
+        /// ```
+        /// let ❰ (foo, bar, .., baz) ❱ = bruh;
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/patterns.html#tuple-patterns)
         struct TuplePat { T!['('], args: [Pat], T![')'] }
 
+        /// Visibility.
+        ///
+        /// ```
+        /// ❰ pub mod ❱ foo;
+        /// ❰ pub(crate) ❱ struct Bar;
+        /// ❰ pub(self) ❱ enum Baz {}
+        /// ❰ pub(super) ❱ fn bruh() {}
+        /// ❰ pub(in bruuh::bruuuh) ❱ type T = u64;
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/visibility-and-privacy.html)
         struct Visibility { T![pub], T![super], T![self], T![crate] }
+
+        /// Single identifier.
+        /// // TODO: clarify the difference between Name and NameRef
+        ///
+        /// ```
+        /// let ❰ foo ❱ = bar;
+        /// struct ❰ Baz ❱;
+        /// fn ❰ bruh ❱() {}
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/identifiers.html)
         struct Name { T![ident] }
-        struct NameRef { /*NameRefToken*/ }
 
-        struct MacroCall: NameOwner, AttrsOwner,DocCommentsOwner {
-            Path, T![!], TokenTree, T![;]
+        /// Reference to a name.
+        ///
+        /// ```
+        /// let foo = ❰ bar ❱(❰ Baz(❰ bruh ❱) ❱;
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/identifiers.html)
+        struct NameRef { /*NameRefToken*/ } // TODO: where is NameRefToken?
+
+        /// Macro call.
+        /// Includes all of its attributes and doc comments.
+        ///
+        /// ```
+        /// ❰
+        ///     /// Docs
+        ///     #[attr]
+        ///     macro_rules! foo {   // macro rules is also a macro call
+        ///         ($bar: tt) => {}
+        ///     }
+        /// ❱
+        ///
+        /// ❰ foo!() ❱
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/macros.html)
+        struct MacroCall: NameOwner, AttrsOwner, DocCommentsOwner {
+            Path, T![!], TokenTree, T![;] // TODO: what is the meaning of the semicolon here?
         }
+
+        /// Attribute.
+        ///
+        /// ```
+        /// ❰ #![inner_attr] ❱
+        ///
+        /// ❰ #[attr] ❱
+        /// ❰ #[foo = "bar"] ❱
+        /// ❰ #[baz(bruh::bruuh = "42")] ❱
+        /// struct Foo;
+        /// ```
+        ///
+        /// [Reference](https://doc.rust-lang.org/reference/attributes.html)
         struct Attr { T![#], T![!], T!['['], Path, T![=], input: AttrInput, T![']'] }
+
+        /// // TODO: clarify on this AST type @edwin0cheng
+        ///
```

Comment:
`TOKEN_TREE` is an ast-node that stores a list of lexer token and other `TOKEN_TREE`. When ra-parser parses an attribute ,  macro_rules and macro call (`foo!), it will convert its contents to a tree of `TokenTree`. e.g :

```
foo!(struct a , use e);
```

will become: 

```
  MACRO_CALL@[0; 28)
    PATH@[0; 3)
      PATH_SEGMENT@[0; 3)
        NAME_REF@[0; 3)
          IDENT@[0; 3) "foo"
    BANG@[3; 4) "!"
    TOKEN_TREE@[4; 27)
      L_PAREN@[4; 5) "("
      STRUCT_KW@[5; 11) "struct"
      WHITESPACE@[11; 12) " "
      IDENT@[12; 13) "a"
      WHITESPACE@[13; 14) " "
      COMMA@[14; 15) ","
      WHITESPACE@[15; 16) " "
      USE_KW@[16; 19) "use"
      WHITESPACE@[19; 20) " "
      TOKEN_TREE@[20; 25)
        L_CURLY@[20; 21) "{"
        WHITESPACE@[21; 22) " "
        IDENT@[22; 23) "e"
        WHITESPACE@[23; 24) " "
        R_CURLY@[24; 25) "}"
      WHITESPACE@[25; 26) " "
      R_PAREN@[26; 27) ")"
    SEMICOLON@[27; 28) ";"
```

Reply:
thanks for a comment added this in 258a3461b428efa0b8cc8fcd40c7844a8ce22406